docs/src/proposals/stake-v2.md

```diff
@@ -0,0 +1,251 @@
+    point_values: [PointValue; 64], // could even go to 512 to match stake history
+}
+```
+
+Where `PointValue` already exists as:
+
+```
+struct PointValue {
+    epoch: Epoch, // epoch concerned
+    rewards: u64, // all new lamports created
+    points: u128, // all points gained
+}
+```
+
+## New sysvar: StakeHistoryV2
+
+During epoch activation, the bank needs to know how much stake is activating,
+effective, and deactivating. Currently, it calculates these by iterating through
+all of the stakes and figuring their activation status based on the existing
+stake history.
+
+For StakeV2, the total information about how much is effective, activating, and
+deactivating must exist somewhere. This can be accomplished through a new sysvar
+that looks just like `StakeHistory`, but only contains information about V2 stakes.
+
+During epoch activation, the bank simply tacks on the V2 information.
+
+Side note: `StakeHistory` still holds all stakes (V1 and V2), so perhaps the
+name `StakeHistoryV2` is incorrect. Suggestions are welcome!
+
+## New Vote Account Fields
+
+Create a new vote account version to store new fields:
+
+ * `withheld_rewards`: all lamports that rightfully belong to delegators, but are stored in the validator vote account
+ * `commission` in `epoch_credits`: to calculate accrued rewards over multiple
+epochs, also store a history of commissions by adding a `u16` to the entries
+in `epoch_credits`.
+
+`Withdraw` needs to respect the withheld amount and rent-exemption. This means
+that vote accounts cannot be deleted if any withheld rewards remain. Thankfully,
+since claiming rewards is permissionless, a node operator can do this themselves
+and then delete their vote account.
+
+Validators may get confused about the lamports in their account, so improve the
+explorer / tools to show how much can be withdrawn, ie
+`total_lamports - rent_exemption - withheld_rewards`.
+
+Side note: while we're making changes to the struct, we can also widen `commission`
+to a `u16` and finally let validators charge 6.9%, see https://degendaoo.academy/dics
+for more context.
+
+## Runtime changes
+
+V2 stakes are not included in the stakes cache, but their delegation amounts are
+added to the vote accounts. Every call to the stake program will still hit the
+cache, but there won't be an entry to process at epoch boundary.
+
+Since the withheld rewards are included in the vote account, it will always have
+a totally accurate delegation amount, regardless of how much is claimed.
+
+Also, since inflation still creates all new lamports at epoch boundary,
+the runtime rules on lamport creation are never broken.
+
+## Downstream effects
+
+### Stake Pools
+
+Stake pools and bots can keep using V1 stake accounts no problem in this model.
+
+Since V2 stakes have redelegation, all pools have a huge incentive to upgrade.
+Currently, during rebalancing, the moved amount loses one epoch of rewards,
+which is a problem for stake pool performance.
```

Comment:
If I'm understanding this correctly, would this mean that stake could be redelegated more than once per epoch in a stake pool using V2 stake accounts? If so, that would be extremely helpful and useful!

Right now, unstaking from a validator in the pool (in circumstances where a validator goes delinquent) results in additional lost rewards since the stake can't start activating to another validator in the same epoch (in one epoch, stake moves from delinquent validator to reserve, in next epoch stake moves from reserve to active validator and begins activation). I think this would mean that two epochs of rewards would be lost whenever unstaking from a validator in the pool (instead of just the one epoch lost for staking newly deposited SOL from reserve to a validator).

Reply:
Correct, this means that you could change your delegation from one validator to another without any lost credits.

Currently, you lose one epoch of rewards while redelegating from validator A to B:
* epoch 1, you deactivate, you still get rewards from delegation to validator A
* epoch 2, you delegate to validator B, no rewards since you're activating
* epoch 3, your delegation is active, and get rewards from delegation to validator B

As a side note, you could use the "redelegate" command more than once per epoch, but the only one that matters for rewards accrual is the chosen validator at the end of the epoch.